File: Experimental/2021/MattsToy2021/src/main/cpp/subsystems/DriveBase.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "subsystems/DriveBase.h"

#include <frc/shuffleboard/Shuffleboard.h>
#include <frc/smartdashboard/SmartDashboard.h>
#include <units/length.h>

#include <wpi/math>

#include "Constants.h"

/// Ticks per revolution on the Rev Neo motors.
constexpr double kTicksPerRevolution_NeoMotor = 42;

/// Gear ratio used for the 2020/2021 robots.
constexpr double kGearRatio_2021 = 10.71;

/// Wheel diameter on the 2020/2021 robots.
/// Makes use of the "units" library provided by WPI.  (See docs at
/// https://docs.wpilib.org/en/stable/docs/software/basic-programming/cpp-units.html.)
static constexpr units::length::inch_t kWheelDiameter = 6.0_in;

static const char* const kShuffleboardTabName = "DriveBase";

DriveBase::DriveBase()
    : m_leftFront(CANBusIds::SparkMax::Left_Front_No,
                  rev::CANSparkMax::MotorType::kBrushless),
      m_leftRear(CANBusIds::SparkMax::Left_Rear_No,
                 rev::CANSparkMax::MotorType::kBrushless),
      m_rightFront(CANBusIds::SparkMax::Right_Front_No,
                   rev::CANSparkMax::MotorType::kBrushless),
      m_rightRear(CANBusIds::SparkMax::Right_Rear_No,
                  rev::CANSparkMax::MotorType::kBrushless),
      m_odometry(units::degree_t(0)) {
  SetSubsystem("Drivebase");

  // The following is why we need to dynamically allocate the speed controller
  // groups and differential drive: we have to mark motors as inverted (or not)
  // *before* putting them into the groups (and then those into the drive); if
  // we do it afterward, it just fails (silently).
  m_leftFront.SetInverted(false);
  m_leftRear.SetInverted(false);
  m_rightFront.SetInverted(true);
  m_rightRear.SetInverted(true);

  leftMotors.reset(new frc::SpeedControllerGroup{m_leftFront, m_leftRear});
  rightMotors.reset(new frc::SpeedControllerGroup{m_rightFront, m_rightRear});

  ConfigureEncoders();

  drive.reset(new frc::DifferentialDrive{*leftMotors, *rightMotors});

  SetCoastingEnabled(false);

  // Gyro must be calibrated to initialize for use (generally immediately on
  // start-up).
  m_adiGyro.Calibrate();

  ConfigureShuffleboard();
}

void DriveBase::ConfigureShuffleboard() {
  auto& tab = frc::Shuffleboard::GetTab(kShuffleboardTabName);
  leftDistance =
      tab.AddPersistent("L Distance", m_leftRearEncoder.GetPosition())
          .GetEntry();
  rightDistance =
      tab.AddPersistent("R Distance", m_rightRearEncoder.GetPosition())
          .GetEntry();
  leftSpeed =
      tab.AddPersistent("L Speed", m_leftRearEncoder.GetVelocity()).GetEntry();
  rightSpeed =
      tab.AddPersistent("R Speed", m_rightRearEncoder.GetVelocity()).GetEntry();
  rotation =
      tab.AddPersistent("Rotation",
                        GetZAxisGyro().GetRotation2d().Degrees().to<double>())
          .GetEntry();
  tab.Add("Reset Odometry", m_resetCommand);
}

void DriveBase::AddToShuffleboard(wpi::StringRef label, frc::Sendable* data,
                                  bool isPersistent) {
  auto& tab = frc::Shuffleboard::GetTab(kShuffleboardTabName);
  if (isPersistent) {
    tab.AddPersistent(label, data);
  } else {
    tab.Add(label, data);
  }
}

void DriveBase::UpdateShuffleboard() {
  const auto wheelSpeeds = GetWheelSpeeds();

  leftDistance.SetDouble(GetLeftDistance().to<double>());
  rightDistance.SetDouble(GetRightDistance().to<double>());
  leftSpeed.SetDouble(wheelSpeeds.left.to<double>());
  rightSpeed.SetDouble(wheelSpeeds.right.to<double>());
  rotation.SetDouble(m_odometry.GetPose().Rotation().Degrees().to<double>());
}

void DriveBase::ConfigureEncoders() {
  // Default for the encoders is to report velocity in RPM; we want that to come
  // back as m/s.
  units::meter_t wheelCircumference =
      kWheelDiameter * wpi::math::pi;  // Will auto-convert to meters
  units::meter_t velocityAdjustmentForGearing =
      wheelCircumference / kGearRatio_2021;  // Should convert RPM to m/min
  units::meter_t velocityAdjustment =
      velocityAdjustmentForGearing / 60;  // Adjust to m/s

  std::cout << "Wheel circumference: " << wheelCircumference
            << "\nVelocity adj. (gearing): " << velocityAdjustmentForGearing
            << "\nVelocity adj. (final): " << velocityAdjustment << std::endl;

  m_leftFrontEncoder.SetVelocityConversionFactor(
      velocityAdjustment.to<double>());
  m_leftRearEncoder.SetVelocityConversionFactor(
      velocityAdjustment.to<double>());
  m_rightFrontEncoder.SetVelocityConversionFactor(
      velocityAdjustment.to<double>());
  m_rightRearEncoder.SetVelocityConversionFactor(
      velocityAdjustment.to<double>());

  ResetEncoders();
}

void DriveBase::UpdateOdometry() {
  auto currentRotation = m_adiGyro.GetRotation2d();
  auto currentLeftDistance = GetLeftDistance();
  auto currentRightDistance = GetRightDistance();

  m_odometry.Update(currentRotation, currentLeftDistance, currentRightDistance);
}

// This method will be called once per scheduler run
void DriveBase::Periodic() {
  // Update odometry data (REQUIRED for trajectory following).
  UpdateOdometry();

  // Update displayed data for debugging/monitoring.
  UpdateShuffleboard();
}

void DriveBase::SetCoastingEnabled(bool enabled) {
  rev::CANSparkMax::IdleMode mode =
      (enabled ? rev::CANSparkMax::IdleMode::kCoast
               : rev::CANSparkMax::IdleMode::kBrake);
  m_leftFront.SetIdleMode(mode);
  m_leftRear.SetIdleMode(mode);
  m_rightFront.SetIdleMode(mode);
  m_rightRear.SetIdleMode(mode);
}

void DriveBase::ArcadeDrive(double xaxisSpeed, double zaxisRotate,
                            bool squareInputs) {
  drive->ArcadeDrive(xaxisSpeed, zaxisRotate, squareInputs);
}

void DriveBase::TankDrive(double leftSpeed, double rightSpeed) {
  std::cout << "Left: " << leftSpeed << ", right: " << rightSpeed << std::endl;
  drive->TankDrive(leftSpeed, -rightSpeed);
}

void DriveBase::ResetEncoders() {
  m_leftFrontEncoder.SetPosition(0.0);
  m_leftRearEncoder.SetPosition(0.0);
  m_rightFrontEncoder.SetPosition(0.0);
  m_rightRearEncoder.SetPosition(0.0);
}

double DriveBase::GetLeftEncoderCount() {
  // For our encoders on Mae/Nike, it's the # of revolutions
  // TODO(mjh): Decide if we want to just use 1 encoder per side, or take an
  // average, or what.
  return m_leftFrontEncoder.GetPosition();
}

double DriveBase::GetRightEncoderCount() {
  // For our encoders on Mae/Nike, it's the # of revolutions
  return m_rightFrontEncoder.GetPosition();
}

units::meter_t DriveBase::GetRightDistance() {
  auto rightDistance = ((m_rightFrontEncoder.GetPosition()) / kGearRatio_2021) *
                       (kWheelDiameter * wpi::math::pi);
  return rightDistance;  // Units will automatically convert
}

units::meter_t DriveBase::GetLeftDistance() {
  auto leftDistance = ((m_leftFrontEncoder.GetPosition()) / kGearRatio_2021) *
                      (kWheelDiameter * wpi::math::pi);
  return leftDistance;  // Units will automatically convert
}

frc::Pose2d DriveBase::GetPose() {
  return m_odometry.GetPose();
}

frc::DifferentialDriveWheelSpeeds DriveBase::GetWheelSpeeds() {
  return frc::DifferentialDriveWheelSpeeds{
      m_leftRearEncoder.GetVelocity() * 1_m / 1_s,
      m_rightRearEncoder.GetVelocity() * 1_m / 1_s};
}

void DriveBase::ResetOdometry(frc::Pose2d pose) {
  ResetEncoders();
  m_odometry.ResetPosition(pose, m_adiGyro.GetRotation2d());
}

void DriveBase::TankDriveVolts(units::volt_t left, units::volt_t right) {
  leftMotors->SetVoltage(left);
  rightMotors->SetVoltage(right);
  drive->Feed();
}

File: Quasics2025Code/MaeTester2025-Java/src/main/java/frc/robot/subsystems/Drivebase.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.subsystems;

import static frc.robot.Constants.CANBusIds.SparkMax.*;

import com.revrobotics.spark.SparkBase.ResetMode;
import com.revrobotics.spark.SparkLowLevel.MotorType;
import com.revrobotics.spark.SparkBase.PersistMode;
import com.revrobotics.spark.config.SparkMaxConfig;
import com.revrobotics.spark.SparkMax;
import com.revrobotics.RelativeEncoder;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.kinematics.DifferentialDriveWheelSpeeds;
import edu.wpi.first.math.util.Units;
import edu.wpi.first.wpilibj.drive.DifferentialDrive;
import edu.wpi.first.wpilibj2.command.SubsystemBase;

public class Drivebase extends SubsystemBase {

  static final double ANDYMARK_6IN_PLACTION_DIAMETER_METERS =
      Units.inchesToMeters(6.0);
  static final double WHEEL_CIRCUMFERENCE_METERS =
      Math.PI * ANDYMARK_6IN_PLACTION_DIAMETER_METERS;
  static final double DRIVEBASE_GEAR_RATIO = 10.71;
  static final double TICKS_PER_REV_FOR_NEO_MOTOR = 42;

  // TODO: We *really* need to replace the use of a motor controller group with
  // CAN-based "leader/follower" configuration.
  final SparkMax m_leftLeader = new SparkMax(LEFT_LEADER_ID, MotorType.kBrushless);
  final SparkMax m_leftFollower = new SparkMax(LEFT_FOLLOWER_ID, MotorType.kBrushless);
  final SparkMax m_rightLeader = new SparkMax(RIGHT_LEADER_ID, MotorType.kBrushless);
  final SparkMax m_rightFollower = new SparkMax(RIGHT_FOLLOWER_ID, MotorType.kBrushless);

  final SparkMaxConfig leftFollowerConfig = new SparkMaxConfig();
  final SparkMaxConfig rightFollowerConfig = new SparkMaxConfig();
  
  private final RelativeEncoder m_leftEncoder;
  private final RelativeEncoder m_rightEncoder;
  private final DifferentialDrive m_drive;

  /** Creates a new Drivebase. */
  public Drivebase() {
    m_rightLeader.setInverted(true);
    m_rightFollower.setInverted(true);

    leftFollowerConfig.follow(LEFT_LEADER_ID);
    rightFollowerConfig.follow(RIGHT_LEADER_ID);

    m_leftFollower.configure(leftFollowerConfig, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters);
    m_rightFollower.configure(rightFollowerConfig, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters);

    m_leftEncoder = m_leftLeader.getEncoder();
    m_rightEncoder = m_rightLeader.getEncoder();
    
    m_drive = new DifferentialDrive(m_leftLeader, m_rightLeader);

    configureEncoders();
  }

  private void configureEncoders() {
    // Default for the encoders is to report distance and velocity in
    // revolutions and rev/minute; we want that in meters and meters/sec.
    final double distanceScalingFactorForGearing =
        WHEEL_CIRCUMFERENCE_METERS / DRIVEBASE_GEAR_RATIO;
    final double velocityScalingFactor = distanceScalingFactorForGearing / 60;

    resetEncoders();
  }

  public void stop() { m_drive.tankDrive(0, 0); }

  public void setMotorSpeed(double leftPercent, double rightPercent) {
    m_drive.tankDrive(leftPercent, rightPercent);
  }

  public void arcadeDrive(double forwardPercent, double turnPercent) {
    m_drive.arcadeDrive(forwardPercent, turnPercent);
  }

  private double getPositionConversionFactor() {
    final double distanceScalingFactorForGearing = WHEEL_CIRCUMFERENCE_METERS / DRIVEBASE_GEAR_RATIO;
    return distanceScalingFactorForGearing;
  }

  private double getVelocityConversionFactor() {
    final double distanceScalingFactorForGearing = WHEEL_CIRCUMFERENCE_METERS / DRIVEBASE_GEAR_RATIO;
    final double velocityScalingFactor = distanceScalingFactorForGearing / 60;
    return velocityScalingFactor;
  }

  public void resetEncoders() {
    m_leftEncoder.setPosition(0);
    m_rightEncoder.setPosition(0);
  }

  public double getLeftEncoderDistanceMeters() {
    return m_leftEncoder.getPosition() * getPositionConversionFactor();
  }
  public double getRightEncoderDistanceMeters() {
    return m_rightEncoder.getPosition() * getPositionConversionFactor();
  }

  public DifferentialDriveWheelSpeeds getWheelSpeeds() {
    return new DifferentialDriveWheelSpeeds(m_leftEncoder.getVelocity() * getVelocityConversionFactor(),
                                            m_rightEncoder.getVelocity() * getVelocityConversionFactor());
  }

  public void tankDriveVolts(double leftVolts, double rightVolts) {
    m_leftLeader.setVoltage(leftVolts);
    m_rightLeader.setVoltage(rightVolts);
  }
}
